Revisiting BBS part 1: removed Q1 and s sig value (#257)
decentralized-identity/bbs-signature: draft-irtf-cfrg-bbs-signatures.md

Procedure:

> 1.  (Q_2, H_1, ..., H_L) = create_generators(L+1)
> 2.  domain = calculate_domain(PK, Q_2, (H_1, ..., H_L), header)
> 3.  if domain is INVALID, return INVALID

Reviewer: ```suggestion
1. (Q_2, H_1, ..., H_L) = create_generators(L+1)
2. domain = calculate_domain(PK, Q_2, (H_1, ..., H_L), header)
3. if domain is INVALID, return INVALID
```
Just fixing the indentation based on my other comment

> 4.  e_octs = serialize((SK, domain, msg_1, ..., msg_L))
> 5.  if e_octs is INVALID, return INVALID
> 6.  e_len = octet_scalar_length
> 7.  e_expand = expand_message(e_octs, expand_dst, e_len)
> 8.  if e_expand is INVALID, return INVALID
> 9.  e = hash_to_scalar(e_expand)
> 11. if e is INVALID, return INVALID
> 12. B = P1 + Q_2 * domain + H_1 * msg_1 + ... + H_L * msg_L
> 13. A = B * (1 / (SK + e))

Reviewer: ```suggestion
4. e = hash_to_scalar(serialize((SK, domain, msg_1, ..., msg_L)))
5. if e is INVALID, return INVALID
6. B = P1 + Q_2 * domain + H_1 * msg_1 + ... + H_L * msg_L
7. A = B * (1 / (SK + e))
```

Not sure if a step checking that `serialize((SK, domain, msg_1, ..., msg_L))` is VALID is needed. If the inputted arguments of Sign are correct, all values passed to `serialize` are scalars that should not cause serialize to abort.

Probably should add `L` in there, but this can be done in another PR. Even if we do though, `serialize` will fail only if `L > 2^64 - 1`, in which case `calculate_domain` would fail at step 2.

> 14. return signature_to_octets(A, e)

Reviewer: ```suggestion
8. return signature_to_octets(A, e)
```
Just fixing the numbering based on my other comment

1.  signature_result = octets_to_signature(signature)
2.  if signature_result is INVALID, return INVALID
3.  (A, e) = signature_result
4.  L = length(messages)
5.  R = length(disclosed_indexes)
6.  U = L - R
7.  (i1, ..., iR) = disclosed_indexes
8.  (j1, ..., jU) = range(1, L) \ disclosed_indexes
9.  (m_1, ..., m_L) = messages
10. (m_i1, ..., m_iR) = (messages[i1], ..., messages[iR])
11. (m_j1, ..., m_jU) = (messages[j1], ..., messages[jU])

Procedure:

1.  (Q_1, MsgGenerators) = create_generators(L+1)
2.  (H_1, ..., H_L) = MsgGenerators
3.  (H_j1, ..., H_jU) = (MsgGenerators[j1], ..., MsgGenerators[jU])
4.  domain = calculate_domain(PK, Q_1, (H_1, ..., H_L), header)
5.  if domain is INVALID, return INVALID
6.  random_scalars = calculate_random_scalars(3+U)
7.  (r1, r2, r3, m~_j1, ..., m~_jU) = random_scalars
8.  B = P1 + Q_1 * domain + H_1 * m_1 + ... + H_L * m_L
9.  Abar = A * r1
10. Bbar = (B - A * e) * r1
11. C = Bbar * r2 + Abar * r3 + H_j1 * m~_j1 + ... + H_jU * m~_jU
12. c = calculate_challenge(Abar, Bbar, C, (i1, ..., iR),
                            (m_i1, ..., m_iR), domain, ph)
13. if c is INVALID, return INVALID

> 14. r4 = r1^-1 (mod r)

Reviewer: ```suggestion
14. r4 = - r1^-1 (mod r)
```

15. r2^ = r2 + r4 * c (mod r)
16. r3^ = r3 + e * r4 * c (mod r)

> 17. for j in (j1, ..., jU): m^_j = m~_j - m_j * c (mod r)

Reviewer: ```suggestion
17. for j in (j1, ..., jU): m^_j = m~_j + m_j * c (mod r)
```

> 17. for j in (j1, ..., jU): m^_j = m~_j + m_j * c (mod r)

Reviewer: ```suggestion
17. for j in (j1, ..., jU): m^_j = m~_j + msg_j * c (mod r)
```

1.  proof_result = octets_to_proof(proof)
2.  if proof_result is INVALID, return INVALID
3.  (Abar, Bbar, c, r2^, r3^, responses) = proof_result
4.  W = octets_to_pubkey(PK)
5.  if W is INVALID, return INVALID
6.  U = length(responses)
7.  R = length(disclosed_indexes)
8.  L = R + U
9.  (i1, ..., iR) = disclosed_indexes
10. (j1, ..., jU) = range(1, L) \ disclosed_indexes

> 11. (m_i1, ..., m_iR) = disclosed_messages
> 12. (m^_j1, ...., m^_jU) = responses

Reviewer: ```suggestion
11. (msg_i1, ..., msg_iR) = disclosed_messages
12. (m^_j1, ...., m^_jU) = commitments
```

Procedure:

1.  (Q_1, MsgGenerators) = create_generators(L+1)
2.  (H_1, ..., H_L) = MsgGenerators
3.  (H_i1, ..., H_iR) = (MsgGenerators[i1], ..., MsgGenerators[iR])
4.  (H_j1, ..., H_jU) = (MsgGenerators[j1], ..., MsgGenerators[jU])

5.  domain = calculate_domain(PK, Q_1, (H_1, ..., H_L), header)
6.  if domain is INVALID, return INVALID

> 7.  D = P1 + Q_1 * domain + H_i1 * m_i1 + ... + H_iR * m_iR

Reviewer: ```suggestion
7.  D = P1 + Q_1 * domain + H_i1 * msg_i1 + ... + H_iR * msg_iR
```

> 8.  C = Bbar * r2^ + Abar * r3^ + H_j1 * m^_j1 + ... + H_jU * m^_jU + D(-c)

Reviewer: ```suggestion
8.  C = Bbar * r2^ + Abar * r3^ + H_j1 * m^_j1 + ... + H_jU * m^_jU + D * c
```

Outputs:

signature, a signature in the form (A, e), where A is a point in G1

>            and e is non-zero scalars mod r.

Reviewer 2: ```suggestion
           and e is a non-zero scalar mod r.
```

### Message Generators

> Following the procedure defined in (#generator-point-computation) with an input count value of 12, for the [BLS12-381-SHAKE-256](#bls12-381-shake-256) suite, outputs the following values (note that the first one corresponds to `Q_1`, while the next 10, to the message generators `H_1, ..., H_10`).

Reviewer 2: Looks like the input count should be 11 now? But that might need changes to the test data.

Author: Good point; pushed a fix. Test data will need an update anyway, but the text will already be correct.

The following section provides an explanation of how the ProofGen and ProofVerify operations work.

Let the prover be in possession of a BBS signature `(A, e)` on messages `msg_1, ..., msg_L` and a `domain` value (see [Sign](#sign)). Let `A = B * (1/(e + SK))` where `SK` the signer's secret key and,
```
B = P1 + Q_1 * domain + H_1 * msg_1 + ... + H_L * msg_L
```
Let `(i1, ..., iR)` be the indexes of generators corresponding to messages the prover wants to disclose and `(j1, ..., jU)` be the indexes corresponding to undisclosed messages (i.e., `(j1, ..., jU) = range(1, L) \ (i1, ..., iR)`). To prove knowledge of a signature on the disclosed messages, work as follows,

- Hide the signature by randomizing it. To randomize the signature `(A, e)`, take uniformly random `r1` in `[1, r-1]`, and calculate,

        1.  A' = A * r1,
        2.  Abar = A' * (-e) + B * r1

>         3.  D = B * r1.

Reviewer 2: This section will need more updates for consistency but I think that's a separate PR
